### PR TITLE
Fix uint32 overflow in CUDA scan _with_indices kernels

### LINKED ISSUE #167086: Unchecked uint32 overflow in CUDA scan (cumsum/cumprod/cummin/etc)
### 🐛 Describe the bug

I observed some incorrect results with `torch.cumsum` when given a tensor with more than 2**32-1 elements.

```
import torch

def max_of_cumsum(x):
    torch.exp(x)
    cumsum = torch.cumsum(x, dim=1)
    return cumsum.max()

x = torch.zeros(
    (309504, 1, 16384),
    device="cuda", dtype=torch.float32)

print(">>> max of cumsum(torch.zeros(...)): %f" % max_of_cumsum(x))

>>> max of cumsum(torch.zeros(...)): 1.000000
```
The expected result is zero. On pytorch 2.9, this code gives a value of 1.0 as the maximum of the cumulative sum of a bunch of zeros.

In `scan_outer_dim` in `ScanUtils.cuh`, we [check](https://github.com/pytorch/pytorch/blob/5c639466f7b1f9453c2a9c0e25b41c3774a12af8/aten/src/ATen/native/cuda/ScanUtils.cuh#L409-L411) that each of three values corresponding to the size of the flattened outer dimensions, the size of the flattened inner dimensions, and the  size of the dimension along which to scan are individually representable as a uint32t, but then [inside the kernel](https://github.com/pytorch/pytorch/blob/5c639466f7b1f9453c2a9c0e25b41c3774a12af8/aten/src/ATen/native/cuda/ScanUtils.cuh#L278), we form the product of the three values in uint32 which can overflow. This causes the kernel to index into the wrong region of memory and apparently pick up an element of the return value of `torch.exp()` in the example above with default CudaCachingAllocator configuration.

### Versions

Tested on Linux with Pytorch 2.9, CUDA 12.8, and an NVIDIA H200 device.

### PR DESCRIPTION
**Pull request by shino16**
Fixes #167086

#167326 added 64-bit index support to `scan_outer_dim` (cumsum/cumprod). This PR does the same for the `_with_indices` variants (cummax/cummin), which still overflow when `num_orows * row_size * num_irows > UINT_MAX`.

Both kernels (`tensor_kernel_scan_outer_dim_with_indices`, `tensor_kernel_scan_innermost_dim_with_indices`) now take an `index_t` template parameter for pointer offset arithmetic. The host functions (`scan_outer_dim_with_indices`, `scan_innermost_dim_with_indices`) check the total element count and dispatch `uint32_t` or `size_t`, matching #167326's pattern. `scan_innermost_dim_with_indices` also widens its `int` locals to `int64_t` and adds `check_fits_in_unsigned` bounds checks that were missing.

## Test plan
`test_cummax_cummin_outer_dim_64bit_indexing` (48GB, CUDA-only): allocates 2^32 int8 elements and verifies `cummax`/`cummin` produce correct results on the outer-dim scan path.

### CODE REVIEW

File: aten/src/ATen/native/cuda/ScanUtils.cuh [hunk 1]
```diff
@@ -206,34 +208,51 @@ __host__ void scan_outer_dim_with_indices(
   dim3 threads(std::min(512, int(num_irows)));
   int64_t maxGridDim = at::cuda::getCurrentDeviceProperties()->maxGridSize[1];
   dim3 grid(std::min(maxGridDim, num_orows), std::min(maxGridDim, ceil_div(num_irows, int64_t{threads.x})));
-  tensor_kernel_scan_outer_dim_with_indices<scalar_t><<<grid, threads, 0, at::cuda::getCurrentCUDAStream()>>>(
-    self.const_data_ptr<scalar_t>(), values.mutable_data_ptr<scalar_t>(), indices.mutable_data_ptr<int64_t>(),
-    num_orows, num_irows, row_size, init, binary_op);
+  if (static_cast<size_t>(num_irows) * num_orows * row_size <= UINT_MAX) {
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/ScanUtils.cuh]: std::numerics_limit here instead?

Comment (Skylion007) [reply aten/src/ATen/native/cuda/ScanUtils.cuh]: Actually we are on CPP20, you can use std::inrange, even cleaner!

Comment (shino16) [reply aten/src/ATen/native/cuda/ScanUtils.cuh]: Thank you! Addressed now in 3364090647e5bb3ba8d7ac111b949344a0273edd.

File: aten/src/ATen/native/cuda/ScanUtils.cuh [hunk 2]
```diff
@@ -206,34 +209,51 @@ __host__ void scan_outer_dim_with_indices(
   dim3 threads(std::min(512, int(num_irows)));
   int64_t maxGridDim = at::cuda::getCurrentDeviceProperties()->maxGridSize[1];
   dim3 grid(std::min(maxGridDim, num_orows), std::min(maxGridDim, ceil_div(num_irows, int64_t{threads.x})));
-  tensor_kernel_scan_outer_dim_with_indices<scalar_t><<<grid, threads, 0, at::cuda::getCurrentCUDAStream()>>>(
-    self.const_data_ptr<scalar_t>(), values.mutable_data_ptr<scalar_t>(), indices.mutable_data_ptr<int64_t>(),
-    num_orows, num_irows, row_size, init, binary_op);
+  if (std::in_range<uint32_t>(static_cast<size_t>(num_irows) * num_orows * row_size)) {
```
[the remark this answers is not in the document]
Comment (shino16) [reply aten/src/ATen/native/cuda/ScanUtils.cuh:230]: As Skylion007 commented [here](https://github.com/pytorch/pytorch/pull/179160#discussion_r3037052497), `std::in_range` is a C++20 API which should be compilable.

File: aten/src/ATen/native/cuda/ScanUtils.cuh [hunk 3]
```diff
@@ -206,34 +209,51 @@ __host__ void scan_outer_dim_with_indices(
   dim3 threads(std::min(512, int(num_irows)));
   int64_t maxGridDim = at::cuda::getCurrentDeviceProperties()->maxGridSize[1];
   dim3 grid(std::min(maxGridDim, num_orows), std::min(maxGridDim, ceil_div(num_irows, int64_t{threads.x})));
```
[the remark this answers is not in the document]
Comment (shino16) [reply aten/src/ATen/native/cuda/ScanUtils.cuh]: `std::min<int64_t>` expects `int64_t` arguments, so there is no cast to `int`.

**Comment by shino16** [comment]
The linter is raising an error:

    >>> General linter failure:
    Error (CLANGTIDY_EXECUTORCH_COMPATIBILITY) Linter failed
      Linter failed. This a bug, please file an issue against the linter
      maintainer.

**Comment by shino16** [comment]
Hi @Skylion007 , this PR should be ready for merge.

**Comment by Skylion007** [comment]
@claude review this PR

**Comment by shino16** [comment]
Thanks! Addressed the review in e1cd03b3c11.

**Comment by Skylion007** [comment]
@claude review this PR specifically the unit tests again.

**Comment by Skylion007** [comment]
@claude review this PR

**Comment by Skylion007** [comment]
@claude review this PR effort:medium
